core/server/proxy/src/main/java/alluxio/proxy/s3/signature/StringToSignProducer.java

```diff
@@ -0,0 +1,317 @@
+  }
+
+  /**
+   * Convert request info to a format string.
+   * @param schema request schema
+   * @param method request method
+   * @param uri request uri
+   * @param signedHeaders signed headers
+   * @param headers request headers
+   * @param queryParams request params
+   * @param unsignedPayload whether or not the payload is unsigned
+   * @return formatted string
+   * @throws S3Exception
+   */
+  public static String buildCanonicalRequest(
+      String schema,
+      String method,
+      String uri,
+      String signedHeaders,
+      Map<String, String> headers,
+      Map<String, String> queryParams,
+      boolean unsignedPayload
+  ) throws S3Exception {
+    String canonicalUri = getCanonicalUri("/", uri);
+
+    String canonicalQueryStr = getQueryParamString(queryParams);
+
+    StringBuilder canonicalHeaders = new StringBuilder();
+
+    for (String header : StringUtils.getStringCollection(signedHeaders, ";")) {
+      canonicalHeaders.append(header.toLowerCase());
+      canonicalHeaders.append(":");
+      if (headers.containsKey(header)) {
+        String headerValue = headers.get(header);
+        canonicalHeaders.append(headerValue);
+        canonicalHeaders.append(NEWLINE);
+
+        // Set for testing purpose only to skip date and host validation.
+        validateSignedHeader(schema, header, headerValue);
+      } else {
+        throw new RuntimeException("Header " + header + " not present in "
+                + "request but requested to be signed.");
+      }
+    }
+
+    String payloadHash;
+    if (UNSIGNED_PAYLOAD.equals(headers.get(X_AMZ_CONTENT_SHA256))
+            || unsignedPayload) {
+      payloadHash = UNSIGNED_PAYLOAD;
+    } else {
+      payloadHash = headers.get(X_AMZ_CONTENT_SHA256);
+    }
+    StringJoiner canonicalRequestJoiner = new StringJoiner(NEWLINE);
+    canonicalRequestJoiner.add(method)
+            .add(canonicalUri)
+            .add(canonicalQueryStr)
+            .add(canonicalHeaders)
+            .add(signedHeaders)
+            .add(payloadHash);
+    return canonicalRequestJoiner.toString();
+  }
+
+  /**
+   * Returns CanonicalUri.
+   *
+   * @param regex Regular expression to split by
+   * @param uri uri string
+   * @return CanonicalUri
+   */
+  private static String getCanonicalUri(String regex, String uri) {
+    StringJoiner result = new StringJoiner(regex);
+    Pattern p = Pattern.compile(regex);
+    Matcher m = p.matcher(uri);
+    int pos = 0;
+    while (m.find()) {
+      result.add(urlEncode(uri.substring(pos, m.start())));
+      pos = m.end();
+    }
+    result.add(urlEncode(uri.substring(pos)));
+    return result.toString();
+  }
+
+  private static String urlEncode(String str) {
+    try {
+      return URLEncoder.encode(str, "UTF-8")
+              .replaceAll("\\+", "%20")
+              .replaceAll("%7E", "~");
+    } catch (UnsupportedEncodingException e) {
+      throw new RuntimeException(e);
+    }
```

Comment:
Can you leave some comments here for the replaceAll?

Reply:
Bump for the above, please explain why those are necessary.